### PR TITLE
New component: FieldValidation

### LINKED ISSUE #234: Create a new `Required` component
It would be great to have a simple react component mark any component as required. It should work similar to the `<Optional>` component we already have in that you pass it a value that needs to exist in order for the condition to be met. 

If the condition is not met, but the block is in the document, the editor should be locked from saving anything using the core APIs. 

```js
const { lockPostSaving, unlockPostSaving } = useDispatch( 'core/editor' );

// we can then lock / unlock the editor using these functions here
lockPostSaving( 'title-empty-lock' );
unlockPostSaving( 'title-empty-lock' );
```

Removing the block should also unlock the editor. 

Additionally, it would be great to show the reason why the editor is locked in a notice when the user tries to save the post & also something in the `PrePublishCheck` to show all the required fields that are empty would be great. 
```[tasklist]
### Tasks
- [ ] Lock the editor when a block containing at least one `Required` component is rendered
- [ ] Unlock the editor when all `Required` components in the entire editor have their conditional met (Each item will handle it's own locked state)
- [ ] Unlock the editor when the block containing the required field gets removed.
```

I think we should be able to use the clean up function of a `useEffect` hook to handle the unlocking on removal for us.

```js
useEffect( () => {

    if (!condition) {
        // lock the editor
    } else {
        // unlock the editor
    }

    return () => {
        // unlock the editor
    }

}, [condition] )
```

### PR DESCRIPTION
**Pull request by darrenjacoby**
### Description

This PR creates a new `<FieldValidation>` component that can be used for required fields or fields that require custom tests for validation.

- Fields can be marked as required.
- Fields can have custom validation tests and responses.
- Editor "update" button is disabled if any field validation is failing on the post or page.

Screenshare with failing field validations;

https://github.com/10up/block-components/assets/15921694/18e43322-2236-46a6-abbd-fad0280a48cc

### Testing

Import `<FieldValidation>` and test with the following;

```js
import { FieldValidation } from '@10up/block-components';

function BlockEdit(props) {
    const { attributes, setAttributes } = props;
    const { title } = attributes;

    return (
        <>
            <FieldValidation
                value={title}
                required={__('Title is a required field', '10up')}
                validate={[
                    [ (value) => /^[a-zA-Z\s]+$/.test(value), __('Title requires alphabetical characters', '10up') ],
                    [ (value) => /^.{10,}$/.test(value), __('Title requires 10 characters or more', '10up') ],
                ]}
            >
                    <RichText
                        tagName="p"
                        value={title}
                        onChange={(value) => setAttributes({ title: value })}
                    />
            </FieldValidation>
        </>
    )
}
```

**Known issue:** `<RichTextCharacterLimit>` also uses floating-ui and seems to impact the ref used by `<FieldValidation>` when nested, impacting the position of validation responses. Solution requires further investigation.

### Changelog
- Added - Field Validation component

### Credits
n/a

### Checklist
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

File: components/field-validation/index.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+ * @param {boolean|Array} validate Field validation validate prop.
+ *
+ * @returns {Array}
+ */
+const createTestsArray = (required, validate) => {
+	const tests = [];
+
+	if (required) {
+		const fn = (str) => typeof str === 'string' && str !== '';
+		const res = typeof required === 'string' ? required : __('Required', '10up-block-library');
```
Comment (fabiankaegy) [inline components/field-validation/index.js]:
```suggestion
		const res = typeof required === 'string' ? required : __('Required', '10up-block-components');
```

**Review by fabiankaegy — CHANGES_REQUESTED** [review]
This is really cool and works like a charm! Thank you for your contribution! 

The only thing I would like to workshop a little bit more is how we style the inline errors. Using floatingUI is great and solves most of our issues already. However I think we should maybe think about making the entire background a solid color so we can ensure legibility. 

@Sidsector9 implemented a similar thing in https://github.com/10up/block-components/pull/235 which we may be able to use as an inspiration. 

Maybe @xavortm also has an idea of what this could look like :)

File: components/field-validation/index.js (hunk 1 above)
Comment (darrenjacoby) [reply components/field-validation/index.js]: Oops, corrected.

**Comment by darrenjacoby** [comment]
Hey @fabiankaegy ,

Thanks for taking a look! There's probably a good case to be made for leaving styling options to developers on each project to account for things like dark-mode/brand colors, etc. But definitely open to design suggestions for defaults.

In the interim, pushed an update which implements an error class on the field itself.

- `.tenup--block-components__validation-error__field` to style the field. For example, adding `background-color: red;` will made the field background red.

image

There are two other available custom classes;

- `.tenup--block-components__validation-error__response` to style the response container.
- `.tenup--block-components__validation-error__response-rule` to style each individual response rule (as there can be multiple rule failures).

**Comment by fabiankaegy** [comment]
I love the addition of the classes. 

However, I kind of disagree that it should be up to the developer to determine how the invalid state should look. The validation issue is not part of the content and counts as part of the editor frame. Having a standardized way to showcase invalid content which then can be overwritten, but in most cases doesn't need to be overwritten, is what I would prefer here. 

The goal of the block components is to remove as much complexity from the developers on individual projects. And having to re-invent how to style the invalid state of a component in the editor doesn't add a lot of value to a client in my perspective. And it creates the risk that when not styled correctly we create very unclear confusing experiences where a user doesn't understand why they are not able to save the post. 

That isn't to say that I don't want anyone to be able to overwrite the default styling. But I think there should be decent opinionated default styling that makes it look as though it were a core API and part of the native editor experience. So using the editors font, color scheme etc.

**Comment by darrenjacoby** [comment]
@fabiankaegy those are good points which all make sense to me! Let’s see how we can implement a nice default state 🙂

**Comment by darrenjacoby** [comment]
I've implemented some basic styling which matches some WordPress defaults. I'll await feedback from [xavortm](https://github.com/xavortm).

Responses;
- Default WordPress font stack
- Font size 13px
- Font weight 500
- WordPress color `var(--wp--preset--color--vivid-red)`

Field;
- Background color of `var(--wp--preset--color--vivid-red)` at 75% opacity.

Example on dark background; 
image

Example on light background;
image

**Comment by xavortm** [comment]
Hi! Joining with some design proposals about the validation UI/UX:

Please see here: https://www.figma.com/file/YtfzsSMXZJb2N2pfnuuAI3/Blocks-validation?type=design&node-id=0%3A1&mode=design&t=HeaSOLhTcoyL21Rc-1

I have listed the steps and states, including singular and combinations blocks (like core/blocks). 

For visual guide/direction, please see the blue section marked as "Ready for dev" on the right. Here is also a screenshot from it:

Screenshot 2023-08-03 at 10 30 20@2x

It uses a component from Gutenberg but also adds a slight shadow to work better on white backgrounds. The Figma docs point to https://developer.wordpress.org/block-editor/how-to-guides/notices/ for the component reference.

**Pull request closed without merging**